docs/proposals/affinity.md

```diff
@@ -0,0 +1,341 @@
+
+For both affinity and anti-affinity, there are two variants: *hard* and
+*soft.* *Hard* rules *must* be satisfied for a pod to schedule, while *soft*
+rules are *hints* that the scheduler will try to satisfy but may not be able to.
+This is similar to the difference between
+[NodeConstraints and NodePreferences](https://github.com/kubernetes/kubernetes/pull/18261).
+
+Note that this proposal is only about *inter-pod* topological affinity and anti-affinity.
+There are other forms of topological affinity and anti-affinity. For example,
+you can use NodeConstraints (NodePreferences) to require (prefer)
+that a set of pods all be scheduled in some zone Z. NodePreferences and NodeConstraints
+are not capable of expressing inter-pod dependencies, and conversely
+the API we descibe here is not capable of expressing NodePreference and NodeConstraint rules.
+For simplicity, we will use the terms "affinity" and "anti-affinity" to mean
+"inter-pod topological affinity" and "inter-pod topological anti-affinity" in the remainder
+of this document.
+
+## API
+
+We will add `AffinitySelector` to `PodSpec`.
+
+```go
+
+AffinitySelector *AffinitySelector `json:"affinitySelector,omitempty"`
+
+type AffinitySelector struct {
+	// When there are multiple elements, the lists of machines corresponding to each
+	// AffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardAffinityExpressions []AffinityTerm `json:"hardAffinityExpression,omitempty"`
+	// When there are multiple elements, the lists of machines corresponding to each
+	// AffinityTerm are intersected, i.e. all terms must be satisfied.
+	HardAntiAffinityExpressions []AffinityTerm `json:"hardAntiAffinityExpression,omitempty"`
+	// Satisfying more expressions is better than satisfying fewer expressions.
+	SoftAffinityExpressions []AffinityTerm `json:"softAffinityExpression,omitempty"`
+	// Satisfying more expressions is better than satisfying fewer expressions.
+	SoftAntiAffinityExpressions []AffinityTerm `json:"softAntiAffinityExpression,omitempty"`
+}
+
+type AffinityTerm struct {
+	PodSelector *PodSelector `json:"podSelector,omitempty"` // empty PodSelector matches all pods
+	TopologyKey string `json:"topologyKey,omitempty"`
+}
+```
+
+One caveat: As noted above, the normal semantics of empty `PodSelector` is to match all pods.
+However, `PodSelector` is scoped to namespace, therefore the normal semantics do not allow
+us to match all pods globally. We need to special-case interpret empty `PodSelector` in this
+context to mean "all pods globally."
+
+To explain how this API works, let's say that the `PodSpec` of a pod `P` has an `AffinitySelector`
+that is configured as follows:
+
+```go
+AffinitySelector {
+	HardAffinityExpressions: {{PodSelector: P1, TopologyKey: "node"}},
+	HardAntiAffinityExpressions: {{PodSelector: P2, TopologyKey: "rack"}},
+	SoftAffinityExpressions: {{PodSelector: P3, TopologyKey: "zone"}},
```

Comment:
Sometimes with soft affinity I would almost prefer a rank expression, I believe this aligns with some of the arguments @quinton-hoole was mentioning.  e.g. - Give me the one closest to data(x).  

/cc @kubernetes/sig-big-data  @kubernetes/rh-storage